### PR TITLE
Cleanup and add make everything functional

### PR DESCRIPTION
**Pull request by worksofliam**
* Seperates logic into units
* Added test cases

### CODE REVIEW

File: src/install.sql [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+create or replace procedure watsonx.SetSpaceIdForJob(spaceid varchar(100))
+begin
+  set watsonx.spaceid = spaceid;
+end;
+
+create or replace procedure watsonx.SetBearerTokenForJob(bearer_token varchar(10000), expires integer)
+modifies sql data
+begin
+  set watsonx.JobBearerToken = bearer_token;
+  set watsonx.JobTokenExpires = current timestamp + expires seconds;
```
Comment (ThePrez) [inline src/install.sql:26]: Need a buffer (recommended minimum 60 seconds) here (gross)

**Review by ThePrez — CHANGES_REQUESTED** [review]

File: src/install.sql (hunk 1 above)
Comment (worksofliam) [reply src/install.sql]: Done

File: src/auth.sql [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  not deterministic
+begin
+  declare expiration_seconds integer;
+  declare needsNewToken char(1) default 'Y';
+  declare successful char(1) default 'N';
+  declare bearer_token       varchar(1400) ccsid 1208;
+
+  set needsNewToken = watsonx.ShouldGetNewToken();
+
+  renew:
```
Comment (forstie) [inline src/auth.sql:13]: what's this for? Sort of looks like what I coded for JG, but its not used in this coding

File: src/auth.sql [hunk 3]
```diff
@@ -0,0 +1,39 @@
+create or replace function watsonx.authenticate() 
+  returns char(1) ccsid 1208
```
Comment (forstie) [inline src/auth.sql:2]: There could be a continue handler or exit handler for unexpected errors that returns N.

File: src/generate.sql [hunk 4]
```diff
@@ -0,0 +1,33 @@
+create or replace function watsonx.generate(
+  text varchar(1000) ccsid 1208,
+  model_id varchar(128) ccsid 1208 default 'meta-llama/llama-2-13b-chat'
+)
+  returns varchar(10000) ccsid 1208
+  not deterministic
```
Comment (forstie) [inline src/generate.sql:6]: no external action
set option usrprf = *user, dynusrprf = *user, commit = *none

File: src/install.sql [hunk 5]
```diff
@@ -0,0 +1,50 @@
+
+create schema watsonx;
+create or replace variable watsonx.apikey varchar(100)  default '';
```
Comment (forstie) [inline src/install.sql:3]: control the ccsid or use for sbcs data

File: src/install.sql [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+create or replace variable watsonx.spaceid varchar(100) default '';
+create or replace variable watsonx.JobBearerToken varchar(10000) default null;
```
Comment (forstie) [inline src/install.sql:5]: control the ccsid or use for sbcs data

File: src/install.sql [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+create or replace variable watsonx.JobTokenExpires timestamp;
+
+--
+-- (possible) TODO: could have a version that uses a database table with row permissions, as mentioned earlier
+--
+
+create or replace procedure watsonx.SetApiKeyForJob(apikey varchar(100))
```
Comment (forstie) [inline src/install.sql:12]: Do you care about performance?

If yes, then change all procedures to use PROGRAM TYPE SUB

File: src/install.sql [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+begin
```
Comment (forstie) [inline src/install.sql:13]: set option usrprf = *user, dynusrprf = *user, commit = *none

File: src/install.sql [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+  set watsonx.apikey = apikey;
+end;
+
+create or replace procedure watsonx.SetSpaceIdForJob(spaceid varchar(100))
+begin
```
Comment (forstie) [inline src/install.sql:18]: set option usrprf = *user, dynusrprf = *user, commit = *none

File: src/install.sql [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+  set watsonx.spaceid = spaceid;
+end;
+
+create or replace procedure watsonx.SetBearerTokenForJob(bearer_token varchar(10000), expires integer)
+modifies sql data
```
Comment (forstie) [inline src/install.sql:23]: set option usrprf = *user, dynusrprf = *user, commit = *none

File: src/install.sql [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+    return 'Y';
+  end if;
+  if (current timestamp > watsonx.JobTokenExpires) then
+    return 'Y';
+  end if;
+  return 'N';
+end;
+
+create or replace procedure watsonx.logoutJob()
+modifies sql data
```
Comment (forstie) [inline src/install.sql:46]: set option usrprf = *user, dynusrprf = *user, commit = *none

File: src/auth.sql (hunk 2 above)
Comment (worksofliam) [reply src/auth.sql:13]: Correct. Not needed.

File: src/generate.sql (hunk 4 above)
Comment (worksofliam) [reply src/generate.sql:6]: Added

File: src/auth.sql [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+  modifies sql data
+  not deterministic
```
[the remark this answers is not in the document]
Comment (worksofliam) [reply src/auth.sql:4]: Done

File: src/install.sql [hunk 13]
```diff
@@ -0,0 +1,50 @@
+
+create schema watsonx;
+create or replace variable watsonx.apikey varchar(100)  default '';
+create or replace variable watsonx.spaceid varchar(100) default '';
```
[the remark this answers is not in the document]
Comment (worksofliam) [reply src/install.sql:4]: Done

File: src/install.sql (hunk 6 above)
Comment (worksofliam) [reply src/install.sql:5]: Done

File: src/install.sql (hunk 8 above)
Comment (worksofliam) [reply src/install.sql:13]: Done

File: src/install.sql (hunk 11 above)
Comment (worksofliam) [reply src/install.sql:46]: Done

**Review by forstie — COMMENTED** [review]
new comments for Liam to consider

File: src/auth.sql [hunk 14]
```diff
@@ -0,0 +1,40 @@
+create or replace function watsonx.authenticate()
```
[the remark this answers is not in the document]
Comment (worksofliam) [reply src/auth.sql:1]: Please make an issue out of this. I agree they're good, but don't belong in this PR.

File: src/install.sql [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+  program type sub
+  set option usrprf = *user, dynusrprf = *user, commit = *none
+begin
+  set watsonx.JobBearerToken = bearer_token;
+  -- We subtract 60 seconds from the expiration time to ensure we don't cut it too close
+  set watsonx.JobTokenExpires = current timestamp + expires seconds - 60 seconds;
+end;
+
+create or replace function watsonx.ShouldGetNewToken() 
+  returns char(1)
```
[the remark this answers is not in the document]
Comment (worksofliam) [reply src/install.sql:37]: Not applicable on this function it seems.

File: src/auth.sql (hunk 14 above)
[the remark this answers is not in the document]
Comment (worksofliam) [reply src/auth.sql:1]: I made an issue! #6

**Review by forstie — APPROVED** [review]
good

**Pull request merged**